src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+        /// Currently, an object is allocated on the LOH if it is larger than 85000 bytes. See LARGE_OBJECT_SIZE in ndp\clr\src\vm\gc.h
+        /// We will go with exactly 80 KBytes, although this is somewhat arbitrary.</summary>
+        private const Int32 MaxShadowBufferSize = 81920;  // Make sure not to get to the Large Object Heap.
+
+        // Removing a private default constructor is a breaking change for the DataContractSerializer.
+        // Because this ctor was here previously we need to keep it around.
+        private BufferedStream() { }
+
+        public BufferedStream(Stream stream)
+
+            : this(stream, _DefaultBufferSize)
+        {
+        }
+
+        public BufferedStream(Stream stream, Int32 bufferSize)
+        {
+            if (stream == null)
+                throw new ArgumentNullException("stream");
+
+            if (bufferSize <= 0)
+                throw new ArgumentOutOfRangeException("bufferSize", SR.Format(SR.ArgumentOutOfRange_MustBePositive, "bufferSize"));
+
+            Contract.EndContractBlock();
+
+            Contract.Assert(!(stream is MemoryStream), "MemoryStream shouldn't be wrapped in a BufferedStream!");
+            Contract.Assert(!(stream is BufferedStream), "BufferedStream shouldn't be wrapped in another BufferedStream!");
+
+            _stream = stream;
+            _bufferSize = bufferSize;
+
+            // Allocate _buffer on its first use - it will not be used if all reads
+            // & writes are greater than or equal to buffer size.
+
+            if (!_stream.CanRead && !_stream.CanWrite)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureNotClosed()
+        {
+            if (_stream == null)
+                throw new ObjectDisposedException(null, SR.ObjectDisposed_StreamClosed);
+        }
+
+        private void EnsureCanSeek()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanSeek)
+                throw new NotSupportedException(SR.NotSupported_UnseekableStream);
+        }
+
+        private void EnsureCanRead()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanRead)
+                throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+        }
+
+        private void EnsureCanWrite()
+        {
+            Contract.Requires(_stream != null);
+
+            if (!_stream.CanWrite)
+                throw new NotSupportedException(SR.NotSupported_UnwritableStream);
+        }
+
+        private void EnsureShadowBufferAllocated()
+        {
+            Contract.Assert(_buffer != null);
+            Contract.Assert(_bufferSize > 0);
+
+            // Already have shadow buffer?
+            if (_buffer.Length != _bufferSize || _bufferSize >= MaxShadowBufferSize)
+                return;
+
+            Byte[] shadowBuffer = new Byte[Math.Min(_bufferSize + _bufferSize, MaxShadowBufferSize)];
```

Comment:
Can it? It looks like this code only gets called when _bufferSize < MaxShadowBufferSize, so the max _bufferSize for this line would be 81919.